feat(app): Connect Robot via IP Address Slidout

koji opened the pull request:
<!--
Thanks for taking the time to open a pull request! Please make sure you've read the "Opening Pull Requests" section of our Contributing Guide:

https://github.com/Opentrons/opentrons/blob/edge/CONTRIBUTING.md#opening-pull-requests

To ensure your code is reviewed quickly and thoroughly, please fill out the sections below to the best of your ability!
-->

# Overview
Added Connect Robot via IP Address slideout to General Settings
This PR closes #8883 
![Screen Shot 2022-03-29 at 12 15 45 AM](https://user-images.githubusercontent.com/474225/160532066-8f2e0615-5d45-4171-8ac7-eb7f32412469.png)

<!--
Use this section to describe your pull-request at a high level. If the PR addresses any open issues, please tag the issues here.
-->

# Changelog
- Added Slideout for adding ip address/hostname
- Added Test to check slideout functionality


<!--
List out the changes to the code in this PR. Please try your best to categorize your changes and describe what has changed and why.

Example changelog:
- Fixed app crash when trying to calibrate an illegal pipette
- Added state to API to track pipette usage
- Updated API docs to mention only two pipettes are supported

IMPORTANT: MAKE SURE ANY BREAKING CHANGES ARE PROPERLY COMMUNICATED
-->

# Review requests
[manually add via ip Figma](https://www.figma.com/file/bVtCogdwiXj3ndpk4AWXCx/Milestone-1%3A-Read-and-Execute?node-id=5318%3A197891)

<!--
Describe any requests for your reviewers here.
-->

# Risk assessment
low

<!--
Carefully go over your pull request and look at the other parts of the codebase it may affect. Look for the possibility, even if you think it's small, that your change may affect some other part of the system - for instance, changing return tip behavior in protocol may also change the behavior of labware calibration.

Identify the other parts of the system your codebase may affect, so that in addition to your own review and testing, other people who may not have the system internalized as much as you can focus their attention and testing there.
-->

koji committed:
Added text from the new design

koji committed:
Added temporary component

Added temporary component to align the functionality with the unified app.

koji committed:
Add ip addr/hostname with existing methods

koji committed:
add 2 text for ip/hostname status

koji committed:
merge edge

koji committed:
Removed styles.css from IpHostnameItem

created a couple of styled components to remove styles.css file

koji committed:
Adding validation to the form

koji committed:
Added validation to the form

- Added validation to the form
- removed copy of IpList that used for a small modification

koji committed:
Call startDiscovery from Refresh btn and Try again btn

koji committed:
Added a state and function for startDiscovery

koji committed:
Added red border for error

koji committed:
Added style to remove the red border

When a user clicks `Add` button, remove the red border from the input box

koji committed:
Updated onClick of the slideout footer button

koji committed:
Removed unused import statements

koji committed:
Added checkIpAndHostname for Refresh and Try again button

koji committed:
Update ManualIpHostname.tsx

koji committed:
Modified style

koji committed:
Modified Add button style

koji committed:
Removed unnecessary tag and style

koji committed:
Added Test cases

koji committed:
Added clicking Add button case

koji committed:
Added availableStatus

koji committed:
update successStatus

koji committed:
remove unused code

koji committed:
Switch from IconButton to Btn+Icon

koji committed:
Added remove ip/hostname case

koji committed:
feat(app): add Connect Robot via IP Address slideout

Add Connect Robot via IP Address slidieout and add test cases for the slideout

fix #8883

koji committed:
Clean up code

- modify slideout height
- removed unused code and comments

koji committed:
Modified import statements

koji committed:
restore the data-testid I removed mistakenly

b-cooper commented on app/src/organisms/AppSettings/ConnectRobotSlideout.tsx:
```diff
@@ -0,0 +1,156 @@
+        {buttonLable}
+      </Link>
+    )
+  }
```
```suggestion
  const displayLinkButton = (buttonLabel: string): JSX.Element => {
    return (
      <Link
        role="button"
        css={TYPOGRAPHY.pSemiBold}
        color={COLORS.blue}
        onClick={props.checkIpAndHostname}
        id="AppSettings_Connection_Button"
      >
        {buttonLabel}
      </Link>
    )
  }
```

b-cooper commented on app/src/organisms/AppSettings/ConnectRobotSlideout.tsx:
```diff
@@ -0,0 +1,156 @@
+const SUPPORT_PAGE_LINK =
+  'https://support.opentrons.com/en/articles/2934336-manually-adding-a-robot-s-ip-address'
+
+interface SP {
```
We've been moving away from using `react-redux`'s `mapStateToProps` and `mapDispatchToProps`  in favor of their hooks api `useSelector` and `useDispatch`.  It cleans up the props interface and tends to be easier to read. 

There are many examples of this usage in our components.  Here's an example file `app/src/organisms/AppSettings/AdvancedSettings.tsx` 

- We import `useSelector` and `useDispatch` from `react-redux`
- Here we would update the props type to just be:
```
interface RobotSlideoutProps {
  isExpanded: boolean
  onCloseClick: () => void
}
```

- then we'd get candidates inside the component function with:
```
const candidates = useSelector((state: State) => 
  getConfig(state)?.discovery.candidates ?? [])
```
- and for dispatch, we'd grab it directly from the hook and make a small internal function for "starting discovery"
```
  const dispatch = useDispatch<Dispatch>()
  const refreshDiscovery = () => dispatch(startDiscovery())
```

b-cooper commented on app/src/organisms/AppSettings/IpHostnameField.tsx:
```diff
@@ -0,0 +1,58 @@
+          {...field}
+          onBlur={event => {
+            field.onBlur(event)
+            if (field.value) submitForm()
```
was this "submit on blur" in the design? I'm not sure I've seen this pattern before in our codebase, just making sure that it's intentional

koji replied:
Answer from Mel.
A user can add an IP address/hostname when the user clicks `Add` button or hits `Enter`.
https://www.figma.com/file/bVtCogdwiXj3ndpk4AWXCx?node-id=5318:197891#170446647

I will remove `onBlur`.

b-cooper commented on app/src/organisms/AppSettings/IpHostnameItem.tsx:
```diff
@@ -0,0 +1,88 @@
+import { Divider } from '../../atoms/structure'
+
+const IpItem = styled.div`
+  flex: 1 1 calc(100% - 4rem);
```
is this `calc` necessary?

b-cooper commented on app/src/organisms/AppSettings/IpHostnameItem.tsx:
```diff
@@ -0,0 +1,88 @@
+          color={COLORS.darkBlack}
+          onClick={remove}
+          marginLeft={SPACING.spacing4}
+          data-testid={`close-button`}
+        >
+          <Icon name={'close'} />
```
```suggestion
          data-testid="close-button"
        >
          <Icon name="close" />
```

b-cooper commented on app/src/organisms/AppSettings/IpHostnameList.tsx:
```diff
@@ -0,0 +1,64 @@
+import type { State } from '../../redux/types'
+import type { DiscoveryCandidates } from '../../redux/config/types'
+import type { Robot, ReachableRobot } from '../../redux/discovery/types'
+
```
Same comment here. Let's use react-redux's hooks api `useSelector` and `useDispatch` for consistency

b-cooper commented on app/src/organisms/AppSettings/ManualIpHostname.tsx:
```diff
@@ -0,0 +1,138 @@
+import type { State } from '../../redux/types'
+import type { DiscoveryCandidates } from '../../redux/config/types'
+
+interface SP {
```
same comment: i.e. prefer `react-redux`'s hooks api over these higher order components

b-cooper commented on app/src/organisms/AppSettings/ManualIpHostname.tsx:
```diff
@@ -0,0 +1,138 @@
+          value={formik.values.ip}
+          height={SPACING.spacing5}
+          width="100%"
+          style={{
```
we should be able to avoid using the inline `style` attribute here by using the `styled(input)` convention

b-cooper commented on app/src/assets/localization/en/app_settings.json:
```diff
@@ -14,7 +14,19 @@
+  "add_ip_hostname": "Add IP Address or Hostname",
+  "connect_ip_button": "Done",
+  "add_ip_button": "Add",
+  "add_ip_error": "Enter a IP Address or Hostname",
```
```suggestion
  "add_ip_error": "Enter an IP Address or Hostname",
```

b-cooper reviewed (approved):
This looks great! 🎊   Congrats on your first technical PR! I've left a few house-keeping comments and suggestions. I checked the functionality and everything works as expected.  The one design thing that I noticed is that the last `IpItem` doesn't have a bottom border in the designs, but one is getting rendered here.  Here's a screenshot of what I'm talking about:

<img width="320" alt="Screen Shot 2022-03-29 at 12 43 16 PM" src="https://user-images.githubusercontent.com/4731953/160663199-680c2644-fbb0-479e-bfe2-50773aab8a4d.png">

b-cooper commented on app/src/organisms/AppSettings/ConnectRobotSlideout.tsx:
```diff
@@ -0,0 +1,156 @@
+      title={t('connect_ip')}
+      onCloseClick={onCloseClick}
+      isExpanded={isExpanded}
+      height={`calc(100vh - ${SPACING.spacing4})`}
```
subtracting out 1rem was to compensate for the height of the BreadCrumbs bar on deeper than tier 1 pages.  This page is a tier 1 route so you don't need to compensate for the height of the non-existent BreadCrumb bar here.
```suggestion
      height="100%"
```

koji committed:
Update app/src/organisms/AppSettings/ConnectRobotSlideout.tsx

Co-authored-by: Brian Arthur Cooper <brian@opentrons.com>

koji committed:
Update app/src/organisms/AppSettings/IpHostnameItem.tsx

Co-authored-by: Brian Arthur Cooper <brian@opentrons.com>

koji committed:
Update app/src/organisms/AppSettings/ConnectRobotSlideout.tsx

Co-authored-by: Brian Arthur Cooper <brian@opentrons.com>

koji committed:
Update app/src/assets/localization/en/app_settings.json

Co-authored-by: Brian Arthur Cooper <brian@opentrons.com>

koji committed:
Removed inline style

koji committed:
Removed a bottom border

koji committed:
Removed mapStateToProps and mapDispatchToProps

switch to useSelector and useDispatch

koji committed:
Removed calc()

koji committed:
Removed onBlur

Details: https://github.com/Opentrons/opentrons/pull/9811#discussion_r837678382

smb2268 commented on app/src/assets/localization/en/app_settings.json:
```diff
@@ -14,7 +14,19 @@
   "update_alerts": "Software Update Alerts",
   "receive_alert": "Receive an alert when an Opentrons software update is available",
-  "connect_ip": "Connect to a Robot via IP Address",
+  "connect_ip": "Connect to Robot via IP Address",
```
This change is causing a test failure in `GeneralSettings.test.tsx`, you can go ahead and update the failing test on line 44 in that file `getByText('Connect to Robot via IP Address')` and it should pass.

koji replied:
Thank you!
I just noticed that the text in GeneralSettings and the title of the slideout are slightly different.
I'm asking about that to Mel. Once I get the answer, I will update.

smb2268 commented on app/src/organisms/AppSettings/IpHostnameItem.tsx:
```diff
@@ -0,0 +1,89 @@
+    <>
+      <Flex justifyContent={JUSTIFY_FLEX_START} alignItems={JUSTIFY_CENTER}>
+        <IpItem>
+          <Text
```
We have a component called `StyledText` in `/atoms/text` that does a lot of this formatting for you. You should be able to replace this `Text` component with `<StyledText as "p">` and get rid of most of these style props! Same goes for the other instances of `Text` in this PR

smb2268 commented on app/src/organisms/AppSettings/ConnectRobotSlideout.tsx:
```diff
@@ -0,0 +1,128 @@
+const SUPPORT_PAGE_LINK =
+  'https://support.opentrons.com/en/articles/2934336-manually-adding-a-robot-s-ip-address'
+
+interface ConnectRobotSlideoutProps {
```
You should export this type so you can import it in your test and cast your props to this type to get rid of the type errors that are failing in CI
```suggestion
export interface ConnectRobotSlideoutProps {
```

smb2268 commented on app/src/organisms/AppSettings/__tests__/ConnectRobotSlideout.test.tsx:
```diff
@@ -0,0 +1,237 @@
+import { i18n } from '../../../i18n'
+import { getScanning, getViewableRobots } from '../../../redux/discovery'
+import { getConfig } from '../../../redux/config'
+import { ConnectRobotSlideoutComponent } from '../ConnectRobotSlideout'
```
```suggestion
import { ConnectRobotSlideout as ConnectRobotSlideoutComponent } from '../ConnectRobotSlideout'
import type { ConnectRobotSlideoutProps } from '../ConnectRobotSlideout'
```

smb2268 commented on app/src/organisms/AppSettings/__tests__/ConnectRobotSlideout.test.tsx:
```diff
@@ -0,0 +1,237 @@
+      checkIpAndHostname: jest.fn(),
+      isExpanded: true,
+      onCloseClick: jest.fn(),
+    }
```
```suggestion
    } as ConnectRobotSlideoutProps
```
If you add this to each instance of props in this file, you should be able to get rid of the current TS errors.

smb2268 reviewed (approved):
This looks great, Koji! I added a few suggestions that should help resolve the failing tests in CI.

koji committed:
Update app/src/organisms/AppSettings/ConnectRobotSlideout.tsx

Co-authored-by: Sarah Breen <sarah.breen@opentrons.com>

koji committed:
Used StyledText instead of Text

koji committed:
Modified test code since updated the components

- updated the component name
- removed duplicated code in the test code
